### PR TITLE
Api 2221

### PR DESCRIPTION
**Pull request by edgarjimenez**
This PR adds the following:
- Filtering of APPLICATION restricted requests
- Ability to cache using custom keys
- Ability to add headers to downstream requests

### CODE REVIEW

File: app/uk/gov/hmrc/apigateway/connector/ServiceConnector.scala [hunk 1]
```diff
@@ -32,5 +32,11 @@ abstract class ServiceConnector(wsClient: WSClient, cache: CacheManager, val ser
   lazy val expiration = getConfInt(s"$serviceName.caching.expirationInSeconds", 120)
 
   override def get[T: ClassTag](urlPath: String)(implicit format: Format[T]): Future[T] =
-    cache.get[T](urlPath, serviceName, super.get(s"$serviceBaseUrl/$urlPath"), caching, expiration)
+    get(urlPath, urlPath)
```
Comment (gokyo) [inline app/uk/gov/hmrc/apigateway/connector/ServiceConnector.scala]: I would change this 
```
override def get[T: ClassTag](urlPath: String)(implicit format: Format[T]): Future[T] =
  get(urlPath, urlPath)
```
to 
```
override def get[T: ClassTag](urlPath: String)(implicit format: Format[T]): Future[T] =
  get(key, urlPath, Seq.empty)
```
The reason is that with the implementation above you perform an extra call in the stack.

File: app/uk/gov/hmrc/apigateway/connector/impl/ThirdPartyApplicationConnector.scala [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+import scala.concurrent.Future
+
+@Singleton
+class ThirdPartyApplicationConnector @Inject() (wsClient: WSClient, cache: CacheManager)
+  extends ServiceConnector(wsClient, cache, "third-party-application") {
+
+  val SERVER_TOKEN_HEADER = "X-server-token"
+
+  def getByServerToken(serverToken: String): Future[Application] = {
+    get[Application](s"$serviceName-$serverToken", s"application", Seq((SERVER_TOKEN_HEADER, serverToken)))
```
Comment (gokyo) [inline app/uk/gov/hmrc/apigateway/connector/impl/ThirdPartyApplicationConnector.scala]: `Seq((SERVER_TOKEN_HEADER, serverToken))` could be replaced with `Seq(SERVER_TOKEN_HEADER -> serverToken)`
In this way we get rid of the extra parentheses

File: app/uk/gov/hmrc/apigateway/play/filter/ApplicationRestrictedEndpointFilter.scala [hunk 3]
```diff
@@ -36,21 +35,30 @@ import scala.concurrent.{ExecutionContext, Future}
   */
 @Singleton
 class ApplicationRestrictedEndpointFilter @Inject()
-(authorityService: AuthorityService)
+(authorityService: AuthorityService, applicationService: ApplicationService)
 (implicit override val mat: Materializer, executionContext: ExecutionContext) extends ApiGatewayFilter {
 
+  def getAppByAuthority(proxyRequest: ProxyRequest): Future[Application] = {
```
Comment (gokyo) [inline app/uk/gov/hmrc/apigateway/play/filter/ApplicationRestrictedEndpointFilter.scala]: `getAppByAuthority` can be private

**Comment by gokyo** [comment]
We need to consider an extra scenario from **API-2221**:


> When I make a request of a user-restricted endpoint with a valid server token
> Then I receive `401 Unauthorized`
> { "code" : "INCORRECT_ACCESS_TOKEN_TYPE", "message" : "The access token type used is not supported when invoking the API." }

This requires us to introduce a new case class in `GatewayError`: `IncorrectAccessTokenType`

File: app/uk/gov/hmrc/apigateway/play/filter/UserRestrictedEndpointFilter.scala [hunk 4]
```diff
@@ -39,20 +38,25 @@ class UserRestrictedEndpointFilter @Inject()
 (authorityService: AuthorityService, scopeValidator: ScopeValidator)
 (implicit override val mat: Materializer, executionContext: ExecutionContext) extends ApiGatewayFilter {
 
+  private def getAuthority(proxyRequest: ProxyRequest) = {
+    authorityService.findAuthority(proxyRequest) recover {
+      case error: NotFound => throw InvalidCredentials()
```
Comment (gokyo) [inline app/uk/gov/hmrc/apigateway/play/filter/UserRestrictedEndpointFilter.scala]: since the variable `error` is not used, I would replace it with `_`

Comment (edgarjimenez) [inline app/uk/gov/hmrc/apigateway/play/filter/UserRestrictedEndpointFilter.scala]: It's not used, but we only throw InvalidCredentials specifically for a NotFound exception.

**Pull request merged**